spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -302,14 +306,31 @@ private UnionReader(List<ValueReader<?>> readers) {
 
     @Override
     public InternalRow read(Decoder decoder, Object reuse) throws IOException {
-      InternalRow struct = new GenericInternalRow(readers.length);
+      // first we need to filter out NULL alternative if it exists in the union schema
+      int nullIndex = -1;
+      List<Schema> alts = schema.getTypes();
+      for (int i = 0; i < alts.size(); i++) {
+        Schema alt = alts.get(i);
+        if (Objects.equals(alt.getType(), Schema.Type.NULL)) {
+          nullIndex = i;
+          break;
+        }
+      }
+      InternalRow struct = new GenericInternalRow(nullIndex >= 0 ? alts.size() - 1 : alts.size());
+      for (int i = 0; i < struct.numFields(); i += 1) {
+        struct.setNullAt(i);
+      }
+
       int index = decoder.readIndex();
       Object value = this.readers[index].read(decoder, reuse);
 
-      for (int i = 0; i < readers.length; i += 1) {
-        struct.setNullAt(i);
+      if (nullIndex < 0) {
+        struct.update(index, value);
+      } else if (index < nullIndex) {
+        struct.update(index, value);
+      } else if (index > nullIndex) {
```

Comment:
Could you add comment on the case when ```index == nullIndex``` here? I think it is handled above but not obvious here given the branch.

Reply:
for `index == nullIndex`, there should be no action, since the returned struct should be all null, and the previous code initialized the struct to be all null, so we can directly return it.